restcomm/restcomm-connect.core/src/main/java/org/restcomm/connect/core/service/profile/ProfileServiceImpl.java

```diff
@@ -0,0 +1,154 @@
+            if (profile == null) {
+                lastAccount = daoManager.getAccountsDao().getAccount(currentAccount);
+                if (lastAccount != null) {
+                    currentAccount = lastAccount.getParentSid();
+                } else {
+                    throw new RuntimeException("account not found!!!");
+                }
+            }
+        } while (profile == null && currentAccount != null);
+
+        // if profile is not found in account hierarchy,try org
+        if (profile == null && lastAccount != null) {
+            Sid organizationSid = lastAccount.getOrganizationSid();
+            profile = retrieveExplicitlyAssociatedProfile(organizationSid.toString());
+        }
+
+        // finally try with default profile
+        if (profile == null) {
+            try {
+                profile = daoManager.getProfilesDao().getProfile(DEFAULT_PROFILE_SID);
+            } catch (SQLException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("Returning profile:" + profile);
+        }
+
+        return profile;
+    }
+
+    /**
+     * @param organizationSid
+     * @return will return associated profile of provided organization sid
+     */
+    public Profile retrieveEffectiveProfileByOrganizationSid(String organizationSid) {
+        Profile profile = null;
+        profile = retrieveExplicitlyAssociatedProfile(organizationSid.toString());
+
+        // finally try with default profile
+        if (profile == null) {
+            try {
+                profile = daoManager.getProfilesDao().getProfile(DEFAULT_PROFILE_SID);
+            } catch (SQLException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+        if (logger.isDebugEnabled()) {
+            logger.debug("Returning profile:" + profile);
+        }
+        return profile;
+    }
+
+    /**
+     * @param targetSid
+     * @return will return associated profile of provided target (account or
+     *         organization)
+     */
+    @Override
+    public Profile retrieveExplicitlyAssociatedProfile(String targetSid) {
+        ProfileAssociation assoc = daoManager.getProfileAssociationsDao().getProfileAssociationByTargetSid(targetSid);
+        Profile profile = null;
+        if (assoc != null) {
+            try {
+                profile = daoManager.getProfilesDao().getProfile(assoc.getProfileSid().toString());
+            } catch (SQLException ex) {
+                throw new RuntimeException("problem retrieving profile", ex);
+            }
+        }
+        if (logger.isDebugEnabled()) {
+            logger.debug("Returning profile:" + profile);
+        }
+        return profile;
+    }
+
+    /**
+     * @param targetSid
+     * @param info
+     * @param resource
+     * @return
+     */
+    @Override
+    public LinkHeader composeProfileLink(String targetSid, UriInfo info, Class resource) {
```

Comment:
not sure whether serviec layer should be aware of REST endpoint stuff, with this we need jaxrs dependency in core lib...